packages/toon/src/encode/encoders.ts

```diff
@@ -356,3 +363,92 @@ function encodeListItemValue(
 }
 
 // #endregion
+
+// #region Delimiter resolution helpers
+
+const AUTO_DELIMITER_PRIORITY: readonly Delimiter[] = [
+  DELIMITERS.tab,
+  DELIMITERS.pipe,
+  DELIMITERS.comma,
+]
+
+function resolveDelimiterForPrimitiveArray(values: readonly JsonPrimitive[], options: ResolvedEncodeOptions): Delimiter {
+  const strings = collectStringsFromPrimitives(values)
+  return selectDelimiter(strings, options)
+}
+
+function resolveDelimiterForArrayOfArrays(values: readonly JsonArray[], options: ResolvedEncodeOptions): Delimiter {
+  const strings: string[] = []
+
+  for (const arr of values) {
+    if (isArrayOfPrimitives(arr)) {
+      collectStringsFromPrimitives(arr, strings)
+    }
+  }
+
+  return selectDelimiter(strings, options)
+}
+
+function resolveDelimiterForTabularArray(rows: readonly JsonObject[], options: ResolvedEncodeOptions): Delimiter {
+  const strings: string[] = []
+
+  for (const row of rows) {
+    for (const value of Object.values(row)) {
+      if (typeof value === 'string') {
+        strings.push(value)
+      }
+    }
+  }
+
+  return selectDelimiter(strings, options)
+}
+
+function collectStringsFromPrimitives(values: readonly JsonPrimitive[], target: string[] = []): string[] {
+  for (const value of values) {
+    if (typeof value === 'string') {
+      target.push(value)
+    }
+  }
+  return target
+}
+
+function selectDelimiter(strings: readonly string[], options: ResolvedEncodeOptions): Delimiter {
+  if (strings.length === 0 || options.delimiterStrategy === 'fixed') {
+    return options.delimiter
+  }
+
+  let bestDelimiter = options.delimiter
+  let bestScore = countDelimiterCollisions(strings, bestDelimiter)
+
+  for (const candidate of AUTO_DELIMITER_PRIORITY) {
+    if (candidate === bestDelimiter) {
+      continue
+    }
+
+    const score = countDelimiterCollisions(strings, candidate)
+    if (score < bestScore) {
+      bestScore = score
+      bestDelimiter = candidate
+
+      if (score === 0) {
+        break
+      }
+    }
+  }
+
+  return bestDelimiter
+}
+
+function countDelimiterCollisions(strings: readonly string[], delimiter: Delimiter): number {
+  let collisions = 0
+
+  for (const value of strings) {
+    if (value.includes(delimiter)) {
+      collisions++
+    }
+  }
+
+  return collisions
+}
```

Comment:
You can simplify this part by using filter to count the items that contain the delimiter.

```suggestion
function countDelimiterCollisions(strings: readonly string[], delimiter: Delimiter): number {
  return strings.filter(s => s.includes(delimiter)).length;
}
```